xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -1733,99 +1708,84 @@ private void sendNackRequest(ResourceType resourceType, Collection<String> resou
       logger.log(XdsLogLevel.DEBUG, "Sent DiscoveryRequest\n{0}", requestProto);
     }
 
     @Override
     void sendError(Exception error) {
-      requestWriterV2.onError(error);
-    }
-  }
-
-  // AdsStream V3
-  private final class AdsStream extends AbstractAdsStream {
-    private final AggregatedDiscoveryServiceGrpc.AggregatedDiscoveryServiceStub stub;
-    private StreamObserver<DiscoveryRequest> requestWriter;
-
-    AdsStream() {
-      stub = AggregatedDiscoveryServiceGrpc.newStub(channel);
+      requestWriter.onError(error);
     }
 
     @Override
-    void start() {
-      StreamObserver<DiscoveryResponse> responseReader = new StreamObserver<DiscoveryResponse>() {
+    public void onNext(final io.envoyproxy.envoy.api.v2.DiscoveryResponse response) {
+      final DiscoveryResponseData responseData = DiscoveryResponseData.fromEnvoyProtoV2(response);
+      if (logger.isLoggable(XdsLogLevel.DEBUG)) {
+        logger.log(XdsLogLevel.DEBUG, "Received {0} response:\n{1}",
+            responseData.getResourceType(), respPrinter.print(response));
+      }
+      syncContext.execute(new Runnable() {
         @Override
-        public void onNext(DiscoveryResponse response) {
-          DiscoveryResponseData responseData =
-              DiscoveryResponseData.fromEnvoyProto(response);
-          if (logger.isLoggable(XdsLogLevel.DEBUG)) {
-            logger.log(
-                XdsLogLevel.DEBUG,
-                "Received {0} response:\n{1}",
-                responseData.getResourceType(),
-                respPrinter.print(response));
-          }
-          onDiscoveryResponse(responseData);
+        public void run() {
+          handleDiscoveryResponse(responseData);
         }
+      });
+    }
 
+    @Override
+    public void onError(final Throwable t) {
+      syncContext.execute(new Runnable() {
         @Override
-        public void onError(Throwable t) {
-          AdsStream.this.onError(t);
+        public void run() {
+          handleStreamClosed(Status.fromThrowable(t));
         }
+      });
+    }
 
+    @Override
+    public void onCompleted() {
+      syncContext.execute(new Runnable() {
         @Override
-        public void onCompleted() {
-          AdsStream.this.onCompleted();
+        public void run() {
+          handleStreamClosed(Status.UNAVAILABLE.withDescription("Closed by server"));
         }
-      };
-      requestWriter = stub.withWaitForReady().streamAggregatedResources(responseReader);
+      });
+    }
+  }
+
+  private final class AdsStreamV3 extends AbstractAdsStream
+      implements StreamObserver<DiscoveryResponse> {
+    private final AggregatedDiscoveryServiceGrpc.AggregatedDiscoveryServiceStub stub;
+    private StreamObserver<DiscoveryRequest> requestWriter;
+
+    AdsStreamV3() {
+      stub = AggregatedDiscoveryServiceGrpc.newStub(channel);
+    }
+
+    @Override
+    void start() {
+      requestWriter = stub.withWaitForReady().streamAggregatedResources(this);
```

Comment:
Here is the only place `LrsStream` is used as a response StreamObserver. `LrsStream` should really be treated as a composition of a request StreamObserver and a response StreamObserver. Not an inheritance of a response StreamObserver.

Reply:
> AdsStream should really be treated as a composition of a request StreamObserver and a response StreamObserver.

Why? Implementing the response observer doesn't break any encapsulation of the class. 

> Not an inheritance of a response StreamObserver.

Again, this is not inheritance, it's adding a contract. "An AdsStream is a StreamObserver" or "An AdsStream does what declared in StreamObserver".

A more extreme case that could possibly make you even more uncomfortable would be [`SerializingExecutor`](https://github.com/grpc/grpc-java/blob/a91acec2d4429c40d3d1fe38745f712f86b0c470/core/src/main/java/io/grpc/internal/SerializingExecutor.java#L36) 😄  But if you think about more deeply, it's natural.

Reply:
`SerializingExecutor` was also discussed before, and there were PRs to make the runnable a field. We were not doing that way just because we wanted to pursue extreme performance for `SerializingExecutor`. (Instantiating a runnable object is an overhead).

Reply:
It's not adding a contract. Who use the contract? It is just because if you stick to call
```
stub.withWaitForReady().streamAggregatedResources(this)
```
then you have to make the whole class as a StreamObserver. Otherwise it's completely not necessary.

Reply:
> I think you also noticed that it's strange it implements Runnable.

A little bit at the first glance. But feel _OK_ and natural when think about it for a couple more seconds. That's a very extreme case. It's a bit abstract, but it is how Java works.

Our discussion is a bit about taste and preference. My point is, there is no harm to do so, not even for readers. But it can simplify and clean up the code a bit, especially for the `AdsStream.this.onError(t);` thing and avoid the somewhat long local class.

Reply:
> It's not adding a contract. Who use the contract?

The stub uses the contract.